dotCMS/src/main/java/com/dotcms/rest/api/v1/secret/ServiceIntegrationHelper.java

```diff
@@ -66,43 +86,65 @@
+            final HttpServletRequest request,
+            final String serviceKey,
+            final PaginationContext paginationContext,
             final User user)
             throws DotSecurityException, DotDataException {
-
         final Optional<ServiceDescriptor> serviceDescriptorOptional = serviceIntegrationAPI
                 .getServiceDescriptor(serviceKey, user);
-        if (serviceDescriptorOptional.isPresent()) {
-            final ImmutableList.Builder<SiteView> hostViewBuilder = new ImmutableList.Builder<>();
-
-            final ServiceDescriptor serviceDescriptor = serviceDescriptorOptional.get();
-            final List<Host> hosts = getHosts(user);
-            final List<Host> hostsWithConfigurations = computeSitesWithConfigurations(serviceKey, hosts, user);
-            for (final Host host : hostsWithConfigurations) {
-                hostViewBuilder.add(new SiteView(host.getIdentifier(), host.getHostname()));
-            }
-            return Optional.of(
-                    new ServiceIntegrationView(serviceDescriptor, hostsWithConfigurations.size(), hostViewBuilder.build())
+        if (!serviceDescriptorOptional.isPresent()) {
+            //Throw exception and allow the mapper do its thing.
+            throw new DoesNotExistException(
+                    String.format("No service integration was found for key `%s`. ", serviceKey)
             );
         }
-        return Optional.empty();
+
+        final OrderDirection orderDirection =
+                UtilMethods.isSet(paginationContext.getDirection()) ? OrderDirection
+                        .valueOf(paginationContext.getDirection()) : OrderDirection.DESC;
+
+        final ServiceDescriptor serviceDescriptor = serviceDescriptorOptional.get();
+        final Map<String,Set<String>> serviceKeysByHost = serviceIntegrationAPI.serviceKeysByHost();
+        final Set<String> sitesWithConfigurations = serviceIntegrationAPI.filterSitesForServiceKey(serviceDescriptor.getKey(), serviceKeysByHost.keySet(), user);
+
+        final PaginationUtil paginationUtil = new PaginationUtil(new SiteViewPaginator(
+                () -> sitesWithConfigurations, hostAPI, contentletAPI));
+        final Response page = paginationUtil
+                .getPage(request, user,
+                        paginationContext.getFilter(),
+                        paginationContext.getPage(),
+                        paginationContext.getPerPage(),
+                        paginationContext.getOrderBy(),
+                        orderDirection,
+                        Collections.emptyMap(),
+                        (Function<PaginatedArrayList<SiteView>, ServiceIntegrationView>) paginatedArrayList -> {
+                            final long count = sitesWithConfigurations.size();
+                            return new ServiceIntegrationView(serviceDescriptor, count,
+                                    paginatedArrayList);
+                        });
+        return page;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider simply returning the value vs storing it in local variable 'page'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4960082)